### PR TITLE
Fix inferred type of async_sessionmaker

### PR DESCRIPTION
**Pull request by Dreamsorcerer**

### CODE REVIEW

File: lib/sqlalchemy/ext/asyncio/session.py [hunk 1]
```diff
@@ -1436,6 +1436,40 @@ async def main():
 
     class_: Type[_AS]
 
+    @overload
+    def __init__(
+        self: async_sessionmaker[AsyncSession],
```
Comment (Dreamsorcerer) [inline lib/sqlalchemy/ext/asyncio/session.py:1441]: This overload does not have `class_`, so mypy will then infer the type as `[AsyncSession]` from the self argument, when no class_ is provided.

File: lib/sqlalchemy/ext/asyncio/session.py [hunk 2]
```diff
@@ -1436,6 +1436,40 @@ async def main():
...
+        expire_on_commit: bool = ...,
+        info: Optional[_InfoType] = ...,
+        **kw: Any,
+    ):
+        ...
+    @overload
+    def __init__(
+        self,
+        bind: Optional[_AsyncSessionBind],
+        class_: Type[_AS],  # type: ignore
```
Comment (Dreamsorcerer) [inline lib/sqlalchemy/ext/asyncio/session.py:1454]: There's a bit of awkwardness here, with the bind parameter being allowed as positional and keyword. So, there's this version without the default value, and the below version with the default value. Not sure if that can be merged into 1 overload somehow...

File: lib/sqlalchemy/ext/asyncio/session.py [hunk 3]
```diff
@@ -1436,6 +1436,40 @@ async def main():
...
+    def __init__(
+        self,
+        bind: Optional[_AsyncSessionBind] = ...,
+        *,
+        class_: Type[_AS],
+        autoflush: bool = ...,
+        expire_on_commit: bool = ...,
+        info: Optional[_InfoType] = ...,
+        **kw: Any,
+    ):
```
Comment (Dreamsorcerer) [inline lib/sqlalchemy/ext/asyncio/session.py:1471]: So, first overload: No class_ provided, self is explicitly `[AsyncSession]`.
Second overload: class_ and bind are provided.
Third overload: class_ is provided as a keyword value, bind is maybe not provided.

**Comment by CaselIT** [comment]
Sadly this does not work. The following code does not type check correctly for mypy
```py
def async_session_factory(
    engine: AsyncEngine,
) -> async_sessionmaker[MyAsyncSession]:
    return async_sessionmaker(engine, class_=MyAsyncSession)


def async_scoped_session_factory(
    engine: AsyncEngine,
) -> async_scoped_session[MyAsyncSession]:
    return async_scoped_session(
        async_sessionmaker(engine, class_=MyAsyncSession),
        scopefunc=lambda: None,
    )
```
```
 2 messages from mypy were not consumed:
/tmp/tmpd4zyqmmw/sessionmakers.py:25: error: Incompatible return value type (got "async_sessionmaker[AsyncSession]", expected "async_sessionmaker[MyAsyncSession]")  [return-value]
/tmp/tmpd4zyqmmw/sessionmakers.py:32: error: Argument 1 to "async_scoped_session" has incompatible type "async_sessionmaker[AsyncSession]"; expected "async_sessionmaker[MyAsyncSession]"  [arg-type]
```

**Comment by Dreamsorcerer** [comment]
Sorry, seems the order matters for some reason I'm not understanding...

**Comment by Dreamsorcerer** [comment]
Give that a try. The self version seems to need to be the last overload.

**Comment by zzzeek** [comment]
seems to work.  please add a test in `test/ext/mypy/plain_files/`

```diff
diff --git a/test/ext/mypy/plain_files/sessionmakers.py b/test/ext/mypy/plain_files/sessionmakers.py
index ce9b766385..0ea5f393ca 100644
--- a/test/ext/mypy/plain_files/sessionmakers.py
+++ b/test/ext/mypy/plain_files/sessionmakers.py
@@ -86,3 +86,15 @@ def main() -> None:
     sess = scoped_fac()
     # EXPECTED_TYPE: MySession
     reveal_type(sess)
+
+
+def test_8837() -> None:
+    as_= async_sessionmaker()
+
+    # EXPECTED_TYPE: async_sessionmaker[AsyncSession]
+    reveal_type(as_)
+
+    async_session = as_()
+
+    # EXPECTED_TYPE: AsyncSession
+    reveal_type(async_session)

```

**Comment by zzzeek** [comment]
oh also sessionmaker

**Comment by zzzeek** [comment]
please add

```diff
diff --git a/lib/sqlalchemy/orm/session.py b/lib/sqlalchemy/orm/session.py
index 849d37ab22..22bcd65db5 100644
--- a/lib/sqlalchemy/orm/session.py
+++ b/lib/sqlalchemy/orm/session.py
@@ -4621,6 +4621,41 @@ class sessionmaker(_SessionClassMethods, Generic[_S]):
 
     class_: Type[_S]
 
+    @overload
+    def __init__(
+        self,
+        bind: Optional[_SessionBind],
+        class_: Type[_S],
+        autoflush: bool = ...,
+        expire_on_commit: bool = ...,
+        info: Optional[_InfoType] = ...,
+        **kw: Any,
+    ):
+        ...
+    @overload
+    def __init__(
+        self,
+        bind: Optional[_SessionBind] = ...,
+        *,
+        class_: Type[_S],
+        autoflush: bool = ...,
+        expire_on_commit: bool = ...,
+        info: Optional[_InfoType] = ...,
+        **kw: Any,
+    ):
+        ...
+    @overload
+    def __init__(
+        self: "sessionmaker[Session]",
+        bind: Optional[_SessionBind] = ...,
+        *,
+        autoflush: bool = ...,
+        expire_on_commit: bool = ...,
+        info: Optional[_InfoType] = ...,
+        **kw: Any,
+    ):
+        ...
+
     def __init__(
         self,
         bind: Optional[_SessionBind] = None,
diff --git a/test/ext/mypy/plain_files/sessionmakers.py b/test/ext/mypy/plain_files/sessionmakers.py
index ce9b766385..c7ef4c0f6a 100644
--- a/test/ext/mypy/plain_files/sessionmakers.py
+++ b/test/ext/mypy/plain_files/sessionmakers.py
@@ -86,3 +86,26 @@ def main() -> None:
     sess = scoped_fac()
     # EXPECTED_TYPE: MySession
     reveal_type(sess)
+
+
+def test_8837_sync() -> None:
+    sm = sessionmaker()
+
+    # EXPECTED_TYPE: sessionmaker[Session]
+    reveal_type(sm)
+
+    session = sm()
+
+    # EXPECTED_TYPE: Session
+    reveal_type(session)
+
+def test_8837_async() -> None:
+    as_= async_sessionmaker()
+
+    # EXPECTED_TYPE: async_sessionmaker[AsyncSession]
+    reveal_type(as_)
+
+    async_session = as_()
+
+    # EXPECTED_TYPE: AsyncSession
+    reveal_type(async_session)
\ No newline at end of file

```

**Comment by Dreamsorcerer** [comment]
Sure, would you also consider tweaking the actual API, as this is a major new release, to only accept keyword arguments after the first argument?

i.e. Implementation would become:
```
    def __init__(
        self,
        bind: Optional[_AsyncSessionBind] = None,
        *,
        class_: Type[_AS] = AsyncSession,  # type: ignore
        autoflush: bool = True,
        expire_on_commit: bool = True,
        info: Optional[_InfoType] = None,
        **kw: Any,
    ):
```

This probably better matches the examples and typical use, and would allow us to drop the first overload that is handling a positional class_.

**Comment by zzzeek** [comment]
> Sure, would you also consider tweaking the actual API, as this is a major new release, to only accept keyword arguments after the first argument?

yes that's a bug that these are not that way right now.    Same for Session, AsyncSession.  I'm kind of amazed these didnt get a "*" in them as I thought i had spent a lot of time finding all these.

**Comment by Dreamsorcerer** [comment]
Should be good now.

**Comment by zzzeek** [comment]
great, can you add the "*" to the plain Session and AsyncSession?

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision fa6d1a8468e98b40e12f82ed7ddaddc1fde060ac of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change fa6d1a8468e98b40e12f82ed7ddaddc1fde060ac: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4224

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

thank you!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4224

**Comment by Dreamsorcerer** [comment]
By the way, I've found this can be done without overloads in future, using `_AS = TypeVar("_AS", bound="AsyncSession", default="AsyncSession")`

Will need `TypeVar` to be imported from typing_extensions for Python <3.12 and mypy implementation hasn't been merged yet, but might be worth updating once it has (and probably waiting for the PEP to be confirmed for 3.12).

**Comment by zzzeek** [comment]
good to know

**Comment by CaselIT** [comment]
oh they finally added support for default in typevar, only 8 release late! The pep is https://peps.python.org/pep-0696/

**Pull request closed without merging**